flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/MemorySegmentWriter.java

```diff
@@ -0,0 +1,191 @@
+            throws MemoryAllocationException {
+        this.serializer = serializer;
+        this.path = path;
+        this.segmentPool = segmentPool;
+        this.outputStream = new ManagedMemoryOutputStream(segmentPool, expectedSize);
+        this.outputView = new DataOutputViewStreamWrapper(outputStream);
+        this.count = 0;
+    }
+
+    @Override
+    public boolean addRecord(T record) throws IOException {
+        if (outputStream.getPos() >= DataCacheWriter.MAX_SEGMENT_SIZE) {
+            return false;
+        }
+        try {
+            serializer.serialize(record, outputView);
+            count++;
+            return true;
+        } catch (IOException e) {
+            if (e.getCause() instanceof MemoryAllocationException) {
+                return false;
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public Optional<Segment> finish() throws IOException {
+        if (count > 0) {
+            return Optional.of(new Segment(path, count, outputStream.getSegments()));
+        } else {
+            segmentPool.returnAll(outputStream.getSegments());
+            return Optional.empty();
+        }
+    }
+
+    /** An output stream subclass that accepts bytes and writes them to memory segments. */
+    private static class ManagedMemoryOutputStream extends OutputStream {
+
+        /** The pool to allocate memory segments from. */
+        private final MemorySegmentPool segmentPool;
+
+        /** The number of bytes in a memory segment. */
+        private final int pageSize;
+
+        /** The memory segments containing written bytes. */
+        private final List<MemorySegment> segments = new ArrayList<>();
+
+        /** The index of the segment that currently accepts written bytes. */
+        private int segmentIndex;
+
+        /** THe number of bytes in the current segment that have been written. */
+        private int segmentOffset;
+
+        /** THe number of bytes that have been written so far. */
+        private long globalOffset;
+
+        public ManagedMemoryOutputStream(MemorySegmentPool segmentPool, long expectedSize)
+                throws MemoryAllocationException {
+            this.segmentPool = segmentPool;
+            this.pageSize = segmentPool.pageSize();
+            this.segmentIndex = 0;
+            this.segmentOffset = 0;
+
+            Preconditions.checkArgument(expectedSize >= 0);
+            ensureCapacity(Math.max(expectedSize, 1L));
+        }
+
+        public long getPos() {
+            return globalOffset;
+        }
+
+        public List<MemorySegment> getSegments() {
+            return segments;
+        }
+
+        @Override
+        public void write(int b) throws IOException {
+            write(new byte[] {(byte) b}, 0, 1);
+        }
+
+        @Override
+        public void write(@Nullable byte[] b, int off, int len) throws IOException {
+            try {
+                ensureCapacity(globalOffset + len);
```

Comment:
Instead of repeatedly allocate `List<MemorySegment>` instance and do divide operation for each write(...), would it be simpler to just maintain `allocatedBytes` and just check `globalOffset + len <= allocatedBytes` for must write() operation?